pkg/policy/l4_filter_test.go

```diff
@@ -117,6 +139,98 @@ func (td *testData) addIdentity(id *identity.Identity) {
 			id.ID: id.LabelArray,
 		}, nil, wg)
 	wg.Wait()
+	td.idSet.Insert(id.ID)
+}
+
+func (td *testData) removeIdentity(id *identity.Identity) {
+	wg := &sync.WaitGroup{}
+	td.sc.UpdateIdentities(
+		nil,
+		identity.IdentityMap{
+			id.ID: id.LabelArray,
+		}, wg)
+	wg.Wait()
+	td.idSet.Remove(id.ID)
+}
+
+func (td *testData) addIdentitySelector(sel api.EndpointSelector) bool {
+	_, added := td.sc.AddIdentitySelector(dummySelectorCacheUser, EmptyStringLabels, sel)
+	return added
+}
+
+func (td *testData) verifyL4PolicyMapEqual(t *testing.T, expected, actual L4PolicyMap, availableIDs ...identity.NumericIdentity) {
+	t.Helper()
+
+	require.Equal(t, expected.Len(), actual.Len())
+	expected.ForEach(func(l4 *L4Filter) bool {
+		port := l4.PortName
+		if len(port) == 0 {
+			port = fmt.Sprintf("%d", l4.Port)
+		}
+
+		l4B := actual.ExactLookup(port, l4.EndPort, string(l4.Protocol))
+		require.NotNil(t, l4B, "Port Protocol lookup failed: [Port: %s, EndPort: %d, Protocol: %s]", port, l4.EndPort, string(l4.Protocol))
+
+		// If no available IDs are provided, we assume the same pointer for
+		// cached selector is used for both expected and actual L4PolicyMap,
+		// just make sure L4 filter is equal
+		if len(availableIDs) == 0 {
+			require.True(t, l4.Equals(l4B), "Expected: %s\nActual: %s", l4.String(), l4B.String())
+			return true
+		}
+
+		require.Equal(t, l4.Port, l4B.Port)
```

Comment:
Why not use `L4Filter.Equals()`?

Reply:
If this is because the PerSelectorPolicies don't match, is there a way you can find to keep them matching?

Reply:
That wouldn't necessarily help us in this case no? For incremental update we still need a custom implementation for checking if the selectors selects the same set of identities based on what is upserted in the selector cache.
`L4Filter.Equals` is a test only function anyways. We can remove the original `Equal` and rely on this custom test equal if we are concerned about the drift in behavior. Diff is also a little bit cleaner with this check implemented in tests.